A2.md

```diff
@@ -0,0 +1,114 @@
+Service Config via DNS
+----------------------
+* Author(s): Mark D. Roth (roth@google.com)
+* Approver: a11r
+* Status: Draft
+* Implemented in: N/A
+* Last updated: 2017-01-19
+* Discussion at: https://groups.google.com/d/topic/grpc-io/DkweyrWEXxU/discussion
+
+## Abstract
+
+This document proposes a mechanism for encoding gRPC service config data
+in DNS for use in the open-source world.
+
+## Background
+
+The [service
+config](https://github.com/grpc/grpc/blob/master/doc/service_config.md)
+mechanism was originally designed for use inside of Google.  However,
+all but one part of the original design will work fine in the open-source
+world.  That one part is the specification of how the service config
+data will be encoded in DNS.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+There are two parts to this proposal.  The first part is to add some
+JSON wrapping for controlling how service config changes are canary
+tested.  The second part describes how the service config is encoded in
+DNS.
+
+### Canarying Changes
+
+When deploying a change to a service config, it is useful to be able to
+canary test changes to avoid wide-spread breakage by slowly increasing the
+number of clients that see the new version.  To that end, multiple
+service configs choices can be listed, in order, along with criteria that
+determine which choice will be selected by a given client:
+
+```
+# A list of one or more service config choices.
+# The first matching entry wins.
+[
+  {
+    # Criteria used to select this choice.
+    # Client language (e.g., 'c++', 'java', 'go', 'python', etc)
+    # Percentage: integer from 0 to 100 indicating the percentage of
+    # clients that should use this choice.
+    'clientLanguage': string,
+    'percentage': number,
+
+    # The service config data object for clients that match the above
+    # criteria.  (The format for this object is defined in
+    # https://github.com/grpc/grpc/blob/master/doc/service_config.md.)
+    'serviceConfig': object
+  }
+]
+```
+
+### Encoding in DNS TXT Records
+
+In DNS, the service config data (in the form documented in the previous
+section) will be encoded in a TXT record with the attribute name
```

Comment:
You should probably call out some references of how this is done for long configs, since [\<character-string\> is limited to 255 bytes](https://tools.ietf.org/html/rfc1035#section-3.3). Specifically, that multiple strings in a single TXT RR are concatenated without space (SPF spec has [most clear description](https://tools.ietf.org/html/rfc4408#section-3.1.3)). I see that mentioned in passing later, but it should be in this section, since this defines the design.

It seems like it may be prudent to reference https://tools.ietf.org/html/rfc1464 as well, either to describe that's the format we're using or explicitly saying we aren't following it.